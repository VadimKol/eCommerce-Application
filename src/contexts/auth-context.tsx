import { createContext, useCallback, useEffect, useMemo, useState } from 'react';

import { apiRoot, RefreshTokenFlow } from '@/api/build-client';
import type { AuthContextInterface } from '@/common/types';
import { broadcastChannel } from '@/common/utils';

export const AuthContext = createContext({
  isAuthenticated: false,
  handleLogin: () => {},
  handleLogout: () => {},
} as AuthContextInterface);

export function AuthProvider({ children }: { children: React.ReactNode }): JSX.Element {
  const [isAuthenticated, setIsAuthenticated] = useState(localStorage.getItem('geek-shop-auth') !== null);

  const handleLogin = useCallback(() => {
    setIsAuthenticated(true);
  }, []);

  const handleLogout = useCallback(() => {
    setIsAuthenticated(false);
  }, []);

  const contextValue = useMemo(
    () => ({ isAuthenticated, handleLogin, handleLogout }),
    [isAuthenticated, handleLogout, handleLogin],
  );

  useEffect(() => {
    const handleAuth = (e: MessageEvent<{ type: string }>): void => {
      if (e.data.type === 'Auth') {
        if (!isAuthenticated) {
          handleLogin();
        } else {
          handleLogout();
        }
      }
      if (e.data.type === 'Password' || e.data.type === 'Auth') {
        Object.assign(apiRoot, RefreshTokenFlow(localStorage.getItem('geek-shop-refresh') || ''));
      }
    };
    broadcastChannel.addEventListener('message', handleAuth);
    return (): void => {
      broadcastChannel.removeEventListener('message', handleAuth);
    };
  }, [isAuthenticated, handleLogin, handleLogout]);

  return <AuthContext.Provider value={contextValue}>{children}</AuthContext.Provider>;
}
