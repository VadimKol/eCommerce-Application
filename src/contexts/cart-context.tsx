import type { Cart, LineItem } from '@commercetools/platform-sdk';
import { createContext, type ReactNode, useCallback, useEffect, useMemo, useState } from 'react';
import { toast } from 'react-toastify';

import { broadcastChannel } from '@/common/utils';

import {
  addPromocodeIntoCart,
  addToCart,
  changeQuantityFromCart,
  clearCart,
  createCart,
  getActiveCart,
  removeFromCart,
} from '../api/cart';

export type CartContextProps = {
  cart: Cart | null;
  isItemInCart: (id: string) => boolean;
  getItemCount: (id: string) => number;
  getLineItemId: (id: string) => string;
  addItemToCart: (id: string, quantity?: number) => Promise<void>;
  removeItemFromCart: (id: string, quantity?: number) => Promise<void>;
  getCartItemsCount: () => number | undefined;
  cartItems: LineItem[];
  updateCart: (updatedCard: Cart | null) => void;
  clearFromCart: (lineItems: string[]) => Promise<void>;
  changeItemQuantityFromCart: (id: string, quantity: number) => Promise<void>;
  addPromocodeToCart: (code: string) => Promise<void>;
};

export const CartContext = createContext({} as CartContextProps);

export function CartProvider({ children }: { children: ReactNode }): React.ReactElement {
  const [cart, setCart] = useState<Cart | null>(null);

  useEffect(() => {
    if (localStorage.getItem('geek-shop-refresh') === null) {
      createCart()
        .then((response) => setCart(response.body))
        .catch(() => toast(`Failed to create cart`, { type: 'error' }));
    } else {
      getActiveCart()
        .then((response) => setCart(response.body))
        .catch(() => toast(`Failed to get cart`, { type: 'error' }));
    }
    const handleCart = (
      e: MessageEvent<{
        type: string;
        payload: { cart: Cart | null };
      }>,
    ): void => {
      if (e.data.type === 'Password' || e.data.type === 'Auth' || e.data.type === 'Cart') {
        setCart(e.data.payload.cart);
      }
    };
    broadcastChannel.addEventListener('message', handleCart);
    return (): void => {
      broadcastChannel.removeEventListener('message', handleCart);
    };
  }, []);

  const updateCart = useCallback((updatedCard: Cart | null): void => {
    setCart(updatedCard);
    broadcastChannel.postMessage({ type: 'Cart', payload: { cart: updatedCard } });
  }, []);

  const cartItems = useMemo(() => (cart ? cart.lineItems : []), [cart]);

  const isItemInCart = useCallback(
    (id: string): boolean => cartItems.some((item) => item.productId === id),
    [cartItems],
  );

  const getLineItemId = useCallback(
    (id: string): string => {
      const item = cartItems.find((cartItem) => cartItem.productId === id);
      return item ? item.id : '';
    },
    [cartItems],
  );

  const getItemCount = useCallback(
    (id: string): number => {
      const item = cartItems.find((cartItem) => cartItem.productId === id);
      return item ? item.quantity : 0;
    },
    [cartItems],
  );

  const addItemToCart = useCallback(
    async (id: string, quantity?: number): Promise<void> => {
      const response = await addToCart(cart!, id, quantity);

      setCart(response.body);
      broadcastChannel.postMessage({ type: 'Cart', payload: { cart: response.body } });
    },
    [cart],
  );

  const removeItemFromCart = useCallback(
    async (id: string, quantity?: number): Promise<void> => {
      const lineItemId = getLineItemId(id);

      if (lineItemId) {
        const response = await removeFromCart(cart!, lineItemId, quantity);

        setCart(response.body);
        broadcastChannel.postMessage({ type: 'Cart', payload: { cart: response.body } });
      }
    },
    [cart, getLineItemId],
  );

  const getCartItemsCount = useCallback(
    (): number => cartItems.reduce((count, item) => count + item.quantity, 0),
    [cartItems],
  );

  const clearFromCart = useCallback(
    async (lineItems: string[]): Promise<void> => {
      const response = await clearCart(cart!, lineItems);

      setCart(response.body);
      broadcastChannel.postMessage({ type: 'Cart', payload: { cart: response.body } });
    },
    [cart],
  );

  const changeItemQuantityFromCart = useCallback(
    async (id: string, quantity: number): Promise<void> => {
      const lineItemId = getLineItemId(id);

      if (lineItemId) {
        const response = await changeQuantityFromCart(cart!, lineItemId, quantity);

        setCart(response.body);
        broadcastChannel.postMessage({ type: 'Cart', payload: { cart: response.body } });
      }
    },
    [cart, getLineItemId],
  );

  const addPromocodeToCart = useCallback(
    async (code: string): Promise<void> => {
      const response = await addPromocodeIntoCart(cart!, code);

      setCart(response.body);
      broadcastChannel.postMessage({ type: 'Cart', payload: { cart: response.body } });
    },
    [cart],
  );

  const CartContextValue: CartContextProps = useMemo(
    () => ({
      cart,
      addItemToCart,
      removeItemFromCart,
      isItemInCart,
      getItemCount,
      getLineItemId,
      cartItems,
      getCartItemsCount,
      updateCart,
      clearFromCart,
      changeItemQuantityFromCart,
      addPromocodeToCart,
    }),
    [
      addItemToCart,
      removeItemFromCart,
      isItemInCart,
      getItemCount,
      getLineItemId,
      cartItems,
      getCartItemsCount,
      cart,
      updateCart,
      clearFromCart,
      changeItemQuantityFromCart,
      addPromocodeToCart,
    ],
  );

  return <CartContext.Provider value={CartContextValue}>{children}</CartContext.Provider>;
}
