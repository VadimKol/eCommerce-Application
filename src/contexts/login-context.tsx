import { createContext, useCallback, useMemo, useState } from 'react';

export interface LoginContextInterface {
  isAuthenticated: boolean;
  setIsAuthenticated: React.Dispatch<React.SetStateAction<boolean>>;
  handleLogout: () => void;
}

export const LoginContext = createContext({
  isAuthenticated: false,
  setIsAuthenticated: () => {},
  handleLogout: () => {},
} as LoginContextInterface);

export function LoginProvider({ children }: { children: React.ReactNode }): JSX.Element {
  const [isAuthenticated, setIsAuthenticated] = useState(false);

  const handleLogout = useCallback(() => {
    setIsAuthenticated(false);
  }, []);

  const contextValue = useMemo(
    () => ({ isAuthenticated, setIsAuthenticated, handleLogout }),
    [isAuthenticated, handleLogout],
  );

  return <LoginContext.Provider value={contextValue}>{children}</LoginContext.Provider>;
}
