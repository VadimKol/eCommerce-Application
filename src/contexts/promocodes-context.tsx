import { createContext, type ReactNode, useEffect, useMemo, useState } from 'react';
import { toast } from 'react-toastify';

import { getPromocodes } from '@/api/cart';
import { DEFAULT_LOCALE } from '@/common/utils';

export interface Promocode {
  id: string;
  name: string;
}

export type PromocodesProps = {
  promocodes: Promocode[] | null;
};

export const PromocodesContext = createContext({} as PromocodesProps);

export function PromocodesProvider({ children }: { children: ReactNode }): React.ReactElement {
  const [promocodes, setPromocodes] = useState<Promocode[] | null>(null);

  useEffect(() => {
    getPromocodes()
      .then((response) =>
        setPromocodes(
          response.body.results.map((promocode) => ({
            id: promocode.id,
            name: promocode.name ? promocode.name[DEFAULT_LOCALE] || '' : '',
          })),
        ),
      )
      .catch(() => toast(`Failed to get promocodes`, { type: 'error' }));
  }, []);

  const CartContextValue: PromocodesProps = useMemo(
    () => ({
      promocodes,
    }),
    [promocodes],
  );

  return <PromocodesContext.Provider value={CartContextValue}>{children}</PromocodesContext.Provider>;
}
