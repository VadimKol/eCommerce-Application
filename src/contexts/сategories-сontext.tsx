import type { ReactNode } from 'react';
import { createContext, useEffect, useMemo, useState } from 'react';
import { useLocation } from 'react-router-dom';

import { getCategories } from '@/api/catalog';
import { NavigationPaths } from '@/common/enums';
import type { CategoriesContextType, CategoriesData } from '@/common/types';
import { StatusError } from '@/common/utils';

export const CategoriesContext = createContext<CategoriesContextType | undefined>(undefined);

type CategoriesProviderProps = {
  children: ReactNode;
};

let categoriesCache: CategoriesData | null = null;

export function CategoriesProvider({ children }: CategoriesProviderProps): JSX.Element {
  const [categories, setCategories] = useState<CategoriesData | null>(categoriesCache);
  const [loading, setLoading] = useState(!categoriesCache);
  const [error, setError] = useState<Error | null>(null);
  const location = useLocation();

  const validatePath = (categoriesData: CategoriesData, pathname: string): void => {
    const pathParts = pathname.split('/').filter(Boolean);

    if (pathParts.length < 2 || `/${pathParts[0]}` !== NavigationPaths.CATALOG.toString()) {
      return;
    }

    const categoryKey = pathParts[1];
    const category = categoriesData.find((catData) => catData.key === categoryKey);

    if (!category) {
      setError(new StatusError('Category not found', 404));
      return;
    }

    if (pathParts.length > 2) {
      const subcategoryKey = pathParts[2];
      const subcategory = category.subcategories.find((subData) => subData.key === subcategoryKey);

      if (!subcategory) {
        setError(new StatusError('Subcategory not found', 404));
      }
    }
  };

  useEffect(() => {
    setError(null);

    if (categoriesCache) {
      setLoading(false);
      setCategories(categoriesCache);
      validatePath(categoriesCache, location.pathname);
      return;
    }

    if (!location.pathname.startsWith(NavigationPaths.CATALOG)) {
      return;
    }

    getCategories()
      .then((data) => {
        categoriesCache = data;
        setCategories(data);
        validatePath(data, location.pathname);
      })
      .catch((err) => {
        if (err instanceof Error) {
          setError(err);
        } else {
          setError(new Error('An unknown error occurred'));
        }
      })
      .finally(() => {
        setLoading(false);
      });
  }, [location.pathname]);

  const value = useMemo(() => ({ categories, loading, error }), [categories, loading, error]);

  return <CategoriesContext.Provider value={value}>{children}</CategoriesContext.Provider>;
}
