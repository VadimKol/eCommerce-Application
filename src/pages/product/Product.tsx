import classNames from 'classnames';
import { useEffect, useState } from 'react';
import { useParams } from 'react-router-dom';

import { getProduct } from '@/api/catalog';
import type { ProductDetails } from '@/common/types';
import { StatusError } from '@/common/utils';
import { Breadcrumbs } from '@/components/breadcrumbs/Breadcrumbs';
import { CategoriesList } from '@/components/categories-list/CategoriesList';
import { useCategories } from '@/hooks/useCategories';
import { NoMatch } from '@/pages/no-match/NoMatch';

import DetailedProduct from './detailed-product/DetailedProduct';
import styles from './styles.module.scss';

export function Product(): JSX.Element {
  const { productName: productKey } = useParams();
  const [product, setProduct] = useState<ProductDetails | null>(null);
  const [loading, setLoading] = useState(true);
  const [subcategoryError, setError] = useState(false);
  const { error: categoryError } = useCategories();

  useEffect(() => {
    const fetchProduct = async (): Promise<void> => {
      try {
        const productData = await getProduct(productKey || '');
        setProduct(productData);
      } catch {
        setError(true);
      } finally {
        setLoading(false);
      }
    };

    void fetchProduct();
  }, [productKey]);

  if (subcategoryError) {
    return <NoMatch />;
  }

  if (categoryError instanceof StatusError && categoryError.statusCode === 404) {
    return <NoMatch />;
  }

  return (
    <main className={classNames('main', styles.main)}>
      <div className={styles.topContainer}>
        <CategoriesList parentClass={styles.categoriesList} />
      </div>
      <Breadcrumbs parentClass={styles.breadcrumbs} />
      <div className={styles.container}>
        <DetailedProduct product={product} loading={loading} />
      </div>
    </main>
  );
}
