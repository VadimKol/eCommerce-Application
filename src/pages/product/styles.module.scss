.main {
  padding-bottom: 20px;
}

.topContainer {
  display: flex;
  justify-content: space-between;
  width: 100%;
  background-image: linear-gradient(to bottom, #085a1f, #088d2e);
}

.categoriesList {
  max-width: var(--max-width);
  margin-inline: auto;
}

.breadcrumbs {
  align-self: flex-start;
  width: fit-content;
  margin-inline: calc(((100% - var(--max-width)) / 2) + 10px);
  margin-block: 10px;

  @media (width <= 1440px) {
    & {
      margin-inline: 10px;
    }
  }
}

.container {
  display: flex;
  flex-direction: column;
  gap: 10px;
  max-width: 100%;
  margin: auto;
  padding: 28px 40px 30px;
  text-align: center;
  color: var(--primary-color);
  background:
    linear-gradient(var(--background-color) 0 0) padding-box,
    linear-gradient(45deg, var(--primary-color), var(--secondary-color)) border-box;
  border: 5px solid transparent;
  border-radius: 10px;
}
