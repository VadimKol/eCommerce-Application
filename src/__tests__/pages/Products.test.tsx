import { render, waitFor } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';

import { CategoriesProvider } from '@/contexts/сategories-сontext';
import { Product } from '@/pages/product/Product';

jest.mock('swiper/css', () => '');
jest.mock('swiper/css/navigation', () => '');
jest.mock('swiper/css/pagination', () => '');
jest.mock('swiper/modules', () => ({
  Navigation: jest.fn(),
  Pagination: jest.fn(),
}));
jest.mock('swiper/react', () => ({
  Swiper: jest.fn(),
  SwiperSlide: jest.fn(),
}));

it('renders correctly', async () => {
  const { container } = render(
    <BrowserRouter>
      <CategoriesProvider>
        <Product />
      </CategoriesProvider>
    </BrowserRouter>,
  );
  await waitFor(() => {
    expect(container).toMatchSnapshot();
  });
});
