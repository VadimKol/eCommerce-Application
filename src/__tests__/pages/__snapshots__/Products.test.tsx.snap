// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div>
  <main
    class="main main"
  >
    <div
      class="topContainer"
    >
      <div
        class="filler"
      >
        Loading categories
      </div>
    </div>
    <div
      class="container"
    >
      <div
        class="infoMessage"
      >
        Loading product data...
      </div>
    </div>
  </main>
</div>
`;
